### PR TITLE
[OPTIMIZER-2.0] REST endpoint to return initial plan

### PR DESCRIPTION
**Pull request by adityamukho**
### Scope & Purpose
Added (undocumented) optional request parameter to the `explain` endpoint to toggle plan optimization.

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/RestHandler/RestExplainHandler.cpp [hunk 1]
```diff
@@ -27,6 +27,7 @@
 #include "Aql/QueryString.h"
 #include "Logger/Logger.h"
 #include "Transaction/StandaloneContext.h"
+#include "Logger/LogMacros.h"
```
Comment (hkernbach) [inline arangod/RestHandler/RestExplainHandler.cpp:30]:
```suggestion
```

File: arangod/RestHandler/RestExplainHandler.cpp [hunk 2]
```diff
@@ -110,7 +111,15 @@ void RestExplainHandler::explainQuery() {
       transaction::StandaloneContext::Create(_vocbase),
       aql::QueryString(queryString), std::move(bindBuilder),
       aql::QueryOptions(optionsSlice));
+
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+  bool optimize = _request->parsedValue("optimize", true);
+
+  LOG_DEVEL << "Optimize: " << std::boolalpha << optimize;
```
Comment (hkernbach) [inline arangod/RestHandler/RestExplainHandler.cpp:118]:
```suggestion
```

File: arangod/RestHandler/RestExplainHandler.cpp [hunk 3]
```diff
@@ -110,7 +110,14 @@ void RestExplainHandler::explainQuery() {
...
       aql::QueryString(queryString), std::move(bindBuilder),
       aql::QueryOptions(optionsSlice));
+
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+  bool optimize = _request->parsedValue("optimize", true);
+
+  auto queryResult = query->explain(optimize);
+#else
   auto queryResult = query->explain();
+#endif
```
Comment (jsteemann) [inline arangod/RestHandler/RestExplainHandler.cpp:120]:
```suggestion
#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
  // optimization can be turned off for debugging purposes only
  bool optimize = _request->parsedValue("optimize", true);
#else
  // optimization cannot be turned off
  constexpr bool optimize = true;
#endif

  auto queryResult = query->explain(optimize);
```

File: arangod/Aql/Query.h [hunk 4]
```diff
@@ -141,7 +141,7 @@ class Query : public QueryContext, public std::enable_shared_from_this<Query> {
   QueryResult parse();
 
   /// @brief explain an AQL query
-  QueryResult explain();
+  QueryResult explain(bool optimize = true);
```
Comment (jsteemann) [inline arangod/Aql/Query.h:144]: I would prefer to not use defaulted parameters, but rather have callers specify the parameter value explicitly.
```suggestion
  QueryResult explain(bool optimize);
```

File: arangod/Aql/Query.cpp [hunk 5]
```diff
@@ -1007,36 +1000,49 @@ QueryResult Query::explain() {
...
-        VPackArrayBuilder guard(result.data.get());
-
-        auto const& plans = opt.getPlans();
-        for (auto& it : plans) {
-          auto& pln = it.first;
-          TRI_ASSERT(pln != nullptr);
-
-          preparePlanForSerialization(pln);
-          pln->toVelocyPack(*result.data.get(), parser.ast(), flags);
+    std::unique_ptr<Optimizer> opt = nullptr;
```
Comment (jsteemann) [inline arangod/Aql/Query.cpp:1003]: nit: doesn't need to be explicitly initialized to nullptr, as this happens automatically.
```suggestion
    std::unique_ptr<Optimizer> opt;
```

**Review by markuspf — APPROVED** [review]

**Pull request merged**